Index: linux-3.10.14/drivers/char/ralink_gpio.c
===================================================================
--- linux-3.10.14.orig/drivers/char/ralink_gpio.c
+++ linux-3.10.14/drivers/char/ralink_gpio.c
@@ -2672,7 +2672,7 @@ irqreturn_t ralink_gpio_irq_handler(int 
 			}
 			else {
 				record[i].rising = now;
-				if (time_before(now, record[i].falling + 200L)) {
+				if (time_before(now, record[i].falling + 1500L)) {
 					//one click
 					schedule_work(&gpio_event_click);
 				}
